### ISSUE TITLE
Intel Compiler compliation fails on QHPR plugin

### ISSUE THREAD

**Issue by nine7nine** [cloudcompare/trunk#434]
Hi,

I have the latest Intel compiler installed and decided to try compiling CC with it. I'm running Archlinux 64bit and the Intel compiler version is;

Intel(R) C Intel(R) 64 Compiler for applications running on Intel(R) 64, Version 17.0.0.098 Build 20160721 Copyright (C) 1985-2016 Intel Corporation.  All rights reserved.
FOR NON-COMMERCIAL USE ONLY

Only after disabling the QHPR plugin was I able to successfully build CC. Output/error; http://pastebin.com/atdVciUb

Intel's compiler also threw a lot of warnings, mostly,  'type qualifier on return type is meaningless'... there were many of them throughout the build. Full output from build; 
http://pastebin.com/YNzr6P6s

Aside from that, the build seems fine. (I've tested some plugins, projects, etc)

thanks

**Comment by cloudcompare** [comment]
That's interesting! Thanks for the feedback.

Apparently the errors are raised by qHull's code... It's a quite old code (2012 I believe). Don't know if there's a new version or known workarounds to make it compile with the Intel compiler...

**Comment by nine7nine** [comment]
Hey, ya, I thought it might be an interesting experiment. I think the Intel-built CC is faster than the GCC one (although my current testing would be unavoidably subjective). I haven't yet experimented too much with compiler optimizations on CC, nor used any actual metrics to test differences. Although some of the plugins seem to do their job quicker...

hmm, maybe I should take a look around, see if there is a newer qhull and/or see if I can get it to compile. I didn't really dig too much. The only other plugins that I didn't test were QGMMREG (i get some non-intel related error on gcc to do with vnl/vector header or something, so I've never had it working) and QKINECT (never have enabled it). Everything else works.

btw, if you have a Mac or Linux box, Intel would probably kick you a non-commercial license for their compiler suite. They offer it to academics, students and people contributing to open source projects. They are pretty slack about it too. I think the big thing is that you can't redistribute binaries, unless holding a different license.

**Comment by ORNis** [comment]
@cloudcompare there are new releases of qHull and we can also replace qHull with CGAL too.

**Comment by nine7nine** [comment]
Well, I had a look at qhull. It looks like the libqhull used in cloudcompare is deprecated. There is a replacement.new version, afaict, called libqhull_r ... I did have a try at replacing qhull, but it looks like there are some changes that I don't know enough about to fix... Not only that, but it complained/error'd on the same line as before. ah well. ;-)

CGAL on the other hand works fine with Intel's Compiler, just compiled v4.9, no problems.

**Comment by ORNis** [comment]
Ok, maybe we can try to replace qhull with CGAL if performance with CGAL are on parity.

**Comment by nine7nine** [comment]
Hey again,

I went ahead and filed a bug report / issue against qhull, regarding the ICC build failure. My linux distribution packages the latest qhull (and provides a source package), so i modified it to use ICC -  caught the build failure and passed it on to them; https://github.com/qhull/qhull/issues/16

Since the older / CC qhull and the current qhull have the same problem AND going under the assumption that CGAL might not work as a replacement for qhull, I figured I would investigate the issue further...Plus, it would be nice to have a fix now ;-) ... So, If they fix it upstream, I'm sure it would be trivial to patch the old version used in CC too. I'll keep an eye on it. thx

**Comment by cloudcompare** [comment]
Great! Thanks for taking care of this task ;)

**Comment by nine7nine** [comment]
No problem! 

I haven't heard back yet, however, I have managed to get QHPR building with ICC. I removed their workaround, that ironically; is for Intel compiler + non-windows qhull. IDK, maybe an older version needed their hack? ...

I patched qhull_a.h && io.c; http://pastebin.com/L3QuAmpN ... then qhull built, without issue. (it works too, as I have tested it). Anyway, we'll wait and see what qhull devs say about it. 

ttyl

**Comment by nine7nine** [comment]
I filed a pull request for your guys review;  https://github.com/cloudcompare/trunk/pull/441

- a commit to get the Intel C/C++ Compiler(17.0) complation fixed on Linux 
- a commit updating libqhull 2012.1=>2015.2 (the last version of libqhull) 

I'm posting here too so that ORNis can also take a look / get a notification. I figure that even if CGAL may replace qhull, it wasn't a big deal for me to update it and it seems to be working fine in CC for me.

As noted in my pull request, I haven't heard back from upstream/qhull, it's been a week, i think. Regardless, i thought since my branch has more than just the ICC fix, it might be worth reviewing. Hopefully, upstream qhull actually gets back to me! ;)

thanks!

**Comment by cloudcompare** [comment]
I confirm that on Windows (Visual 2013), qHull 2015 works like a charm.

**Comment by nine7nine** [comment]
beautiful! ... I didn't think there would be a problem. ;-)

**Comment by ORNis** [comment]
I close this for now, since the main issue is solved. I think the CGAL vs qHull in CC is something that should be discussed somewhere else.  Thank you for your valuable contribution.
